### ISSUE TITLE
[Regression] 0 bytes processed in adapter response

### ISSUE THREAD

**Issue by abuckenheimer** [dbt-labs/dbt-bigquery#602]
### Is this a regression in a recent version of dbt-bigquery?

- [X] I believe this is a regression in dbt-bigquery functionality
- [X] I have searched the existing issues, and I could not find an existing issue for this regression

### Current Behavior

Hey folks this is a bit of a convoluted "regression" so I'm happy to mark this as something else but following the merge of #77 I noticed that the `bytes_processed` returned in the adapter response for models with `copy_partitions: true` is now 0. This kind of makes sense, bigquery copies are free so 0 bytes is kind of correct but its ultimately unexpected because the construction of the temporary table is ultimately not free.

This is indicative of a larger problem with the adapter response in that it only returns the bytes processed for the last step, it does not consider every part of running the model. You can see this when `copy_partitions: false`, dbt is under reporting the cost by up to half of running incremental models because we big-query will charge you for both the creation of a temporary table and then again to read the temporary table for the merge query into the destination.

### Expected/Previous Behavior

given a similar config to:
```
model_name:
    +materialized: incremental
    +partition_by:
              copy_partitions: 
              field: timestamp
              data_type: timestamp
              granularity: day
```

On a dbt run I'd expect to see the bytes processed be equal to the amount of bytes that would be processed by running the `target/compiled/.../model_name.sql` file in the bigquery console, instead I see:
```
19:20:57  1 of 1 OK created sql incremental model dataset.model_name................ [None (0 processed) in 166.72s]
```

### Steps To Reproduce

see above

### Environment

```markdown
- OS:Ubuntu 18.04.6 LTS (Bionic Beaver)
- Python: 3.11.2
- dbt-core (working version): N/A
- dbt-bigquery (working version): N/A 
- dbt-core (regression version): 1.5.0-b3
- dbt-bigquery (regression version): 1.5.0b2
```

**Comment by dbeatty10** [comment]
Thanks for opening this @abuckenheimer !

Agree with your assessment that this is because the adapter response only returns the bytes processed for the last step, and it does not consider each query that executed as a result of running the model. This is problematic when it appears to the user that it is counting everything.

Although this may be surprising to many dbt-bigquery users, the current behavior is what we expect, so I relabeled this as a feature request / enhancement. I also re-titled accordingly.

We could imagine some scenario where dbt was able to accumulate the bytes processed over multiple statements and then display them. This could be a useful proxy for how much it cost to run a particular model.

There's a few main things that give me pause though:
1. The output emitted by dbt doesn't feel like the best place for comprehensive cost tracking and management
1. Not all cost-generating activity is likely to be easily captured and assigned to an individual dbt model -- think API calls or necessary metadata queries.
1. We typically prefer to expose features that are supported on multiple data platforms (e.g., not just a single platform like BigQuery)

`bytes_processed` in dbt-bigquery is already an exception to the norm of only exposing features supported by multiple warehouses, and we could push further down that path by trying to do an accumulation / roll-up of the reported value.

But my dominating thought is that most cost is going to be incurred once a project is in production, and it seems crucial that teams have a method to monitor their costs outside of the CLI output of dbt. I'd rather analytics engineers be comfortable seeing their all-in costs (including consumption by end users like BI) than relying on the reported bytes during model development.

Interested to hear your feedback -- I am overestimating the role of comprehensive cost observability (outside of the dbt output)?

**Comment by abuckenheimer** [comment]
I agree I don't expect dbt console reporting to be comprehensive but its the most practical feedback loop you can have while developing. We ask our developers to come up with an estimate of cost in PR's when introducing new models or making changes and seeing the gb processed in the console is a really great starting point telling us is this going up if so by how much. We know this is just an estimate and misses the nuance you might get in running a model over a period of time and seeing client read patterns but at least it tells you within an order of magnitude what something might cost which is most of what matters in evaluating changes. Now that this just reports 0 you can't even start a discussion without bringing in other tools.

**Comment by dbeatty10** [comment]
I hear you about supporting a practical feedback loop and how the optimization in https://github.com/dbt-labs/dbt-bigquery/issues/77 removed some information that was there previously (however imperfect it was).

Although we aren't able to prioritize it ourselves at this time, we'd welcome a PR from a community member that provides a targeted, best-effort roll-up of bytes processed on a per-model basis. Labeling this accordingly as `help_wanted`.

**Comment by tanghyd** [comment]
> Agree with your assessment that this is because the adapter response only returns the bytes processed for the last step, and it does not consider each query that executed as a result of running the model. This is problematic when it appears to the user that it is counting everything.

Is this fact reported anywhere in the `dbt-bigquery` documentation? If it is documented outside of this issue and I've simply missed it, please ignore this comment.

Anyway, this issue caught us out recently as an intermediate temporary table step was incurring a bytes billed charge that was a significant order of magnitude (e.g. placeholder numbers like >$5 instead of <$0.005) greater than the bytes billed reported by the CLI. Even on a frequently refreshed incremental table, this can add up significantly.

> Now that this just reports 0 you can't even start a discussion without bringing in other tools.

👍  Like @abuckenheimer, we would typically report approximate costs in PRs (based on the CLI, or more specifically the adapter response metadata attached to the returned `dbtRunnerResult` object) but because of this hidden temporary table creation cost it was missed during our review process.

Most of the time this has been fine until now when it costed us $$$ as a surprise christmas present 😅. Thankfully we managed to catch the full bytes billed via a completely different monitoring system and stopped the issue before it got out of hand. However, if we knew this metric from the dbt-bigquery adapter was not reliable in the first place we would not have built around it assuming it was.

> Although we aren't able to prioritize it ourselves at this time, we'd welcome a PR from a community member that provides a targeted, best-effort roll-up of bytes processed on a per-model basis. Labeling this accordingly as help_wanted.

If I can find some time to code up a solution to this issue I will, but in the absence of a PR to solve this I think this needs to be documented somewhere more visible for the average user of dbt-bigquery.

**Comment by dbeatty10** [comment]
These two issues also asked for all bytes processed to be included in the adapter response and `run_results.json`:
- https://github.com/dbt-labs/dbt-bigquery/issues/775
- https://github.com/dbt-labs/dbt-bigquery/issues/1230

I closed them each in favor of this issue.

**Comment by richardbiddle** [comment]
This is an earlier issue report which is related to this request: https://github.com/dbt-labs/dbt-bigquery/issues/293

**Comment by codigo-ergo-sum** [comment]
I don't have the opportunity to submit a PR right now for this, but just chiming in as I am the original poster/opener for #775. Understanding the practical challenges of doing so, I am still very interested in seeing this functionality included.
